cpp/src/components/weakly_connected_components.cu

```diff
@@ -0,0 +1,741 @@
+    static_cast<vertex_type *>(nullptr), static_cast<vertex_type *>(nullptr)))) edge_buffer_first{};
+  // FIXME: we can use cuda::atomic instead but currently on a system with x86 + GPU, this requires
+  // placing the atomic barrier on managed memory and this adds additional complication.
+  size_t *num_edge_inserts{};
+  size_t next_bucket_idx{};
+
+  template <bool multi_gpu = GraphViewType::is_multi_gpu>
+  __device__ std::enable_if_t<multi_gpu, thrust::optional<thrust::tuple<size_t, std::byte>>>
+  operator()(thrust::tuple<vertex_type, vertex_type> tagged_v, int v_val /* dummy */) const
+  {
+    auto tag = thrust::get<1>(tagged_v);
+    auto v_offset =
+      vertex_partition.get_local_vertex_offset_from_vertex_nocheck(thrust::get<0>(tagged_v));
+    // FIXME: better switch to atomic_ref after
+    // https://github.com/nvidia/libcudacxx/milestone/2
+    auto old =
+      atomicCAS(level_components + v_offset, invalid_component_id<vertex_type>::value, tag);
+    if (old != invalid_component_id<vertex_type>::value && old != tag) {  // conflict
+      static_assert(sizeof(unsigned long long int) == sizeof(size_t));
+      auto edge_idx = atomicAdd(reinterpret_cast<unsigned long long int *>(num_edge_inserts),
+                                static_cast<unsigned long long int>(1));
+      *(edge_buffer_first + edge_idx) = thrust::make_tuple(tag, old);
+    }
+    return (old == invalid_component_id<vertex_type>::value)
+             ? thrust::optional<thrust::tuple<size_t, std::byte>>{thrust::make_tuple(
+                 next_bucket_idx, std::byte{0} /* dummy */)}
+             : thrust::nullopt;
+  }
+
+  template <bool multi_gpu = GraphViewType::is_multi_gpu>
+  __device__ std::enable_if_t<!multi_gpu, thrust::optional<thrust::tuple<size_t, std::byte>>>
+  operator()(thrust::tuple<vertex_type, vertex_type> tagged_v, int v_val /* dummy */) const
+  {
+    return thrust::optional<thrust::tuple<size_t, std::byte>>{
+      thrust::make_tuple(next_bucket_idx, std::byte{0} /* dummy */)};
+  }
+};
+
+template <typename GraphViewType>
+void weakly_connected_components_impl(raft::handle_t const &handle,
+                                      GraphViewType const &push_graph_view,
+                                      typename GraphViewType::vertex_type *components,
+                                      bool do_expensive_check)
+{
+  using vertex_t = typename GraphViewType::vertex_type;
+  using edge_t   = typename GraphViewType::edge_type;
+  using weight_t = typename GraphViewType::weight_type;
+
+  static_assert(std::is_integral<vertex_t>::value,
+                "GraphViewType::vertex_type should be integral.");
+  static_assert(!GraphViewType::is_adj_matrix_transposed,
+                "GraphViewType should support the push model.");
+
+  auto const num_vertices = push_graph_view.get_number_of_vertices();
+  if (num_vertices == 0) { return; }
+
+  // 1. check input arguments
+
+  CUGRAPH_EXPECTS(
```

Comment:
Should this check be here, or would it be better up a level?

Thought being... SCC can be implemented as calling WCC on a directed graph, and then reversing the edges (transposing the matrix) and calling a customized WCC on the reversed/transposed graph.  Pushing this check up to the weakly_connected_components function would make that easier.

Of course, we can always do that when we get around to SCC.

Reply:
> Should this check be here, or would it be better up a level?
> 
Doesn't matter much, this can be here or one level up. But my rational for placing this check here is to avoid the upper level function simple (which is just a caller to the actual implementation function). Based on algorithms, checking for input arguments (especially when do_expensive_check is enabled) can become pretty lengthy.

And I haven't thought much about the behavior of this function if the input graph is asymmetric... or using this WCC as part of SCC implementation. I think we can revisit this later. Designing efficient SCC algorithm requires some additional thinking.